core/src/main/scala/zio/jdbc/SqlFragment.scala

```diff
@@ -373,16 +372,47 @@ object SqlFragment {
     implicit val bigIntSetter: Setter[java.math.BigInteger]           = bigDecimalSetter.contramap(new java.math.BigDecimal(_))
     implicit val bigDecimalScalaSetter: Setter[scala.math.BigDecimal] = bigDecimalSetter.contramap(_.bigDecimal)
     implicit val byteChunkSetter: Setter[Chunk[Byte]]                 = byteArraySetter.contramap(_.toArray)
-    implicit val instantSetter: Setter[java.time.Instant]             = sqlTimestampSetter.contramap(java.sql.Timestamp.from)
+
+    // These `java.time.*` are inspired from Quill encoders. See `ObjectGenericTimeEncoders` in Quill.
+    // Notes:
+    //   1. These setters probably don't work for SQLite. Quill as a separate trait, named `BasicTimeDecoders` which seems dedicated to SQLite.
+    //   2. We deliberately decided not to support `java.time.OffsetTime`.
+    //      Because:
+    //        - See https://github.com/h2database/h2database/issues/521#issuecomment-333517705
+    //        - It's supposed to be mapped to `java.sql.Types.TIME_WITH_TIMEZONE` but this type isn't supported by the PG JDBC driver.
+    //          See: https://github.com/pgjdbc/pgjdbc/blob/9cf9f36a1d3a1edd9286721f9c0b9cfa9e8422e3/pgjdbc/src/main/java/org/postgresql/jdbc/PgPreparedStatement.java#L557-L741
+    //      Note that Quill made a different choice. For PG, it uses `java.sql.Types.TIME` but as we don't support yet differences between DBs and `OffsetTime` is almost never used
+    //      it's simpler, for now, to not support it and to document this choice.
+    //      If you need it, please open an issue or a PR explaining your use case.
+    implicit val sqlDateSetter: Setter[java.sql.Date]                   = forSqlType((ps, i, value) => ps.setDate(i, value), Types.DATE)
+    implicit val sqlTimeSetter: Setter[java.sql.Time]                   = forSqlType((ps, i, value) => ps.setTime(i, value), Types.TIME)
+    implicit val sqlTimestampSetter: Setter[java.sql.Timestamp]         =
+      forSqlType((ps, i, value) => ps.setTimestamp(i, value), Types.TIMESTAMP)
+    implicit val localDateSetter: Setter[java.time.LocalDate]           =
+      sqlDateSetter.contramap(java.sql.Date.valueOf)
+    implicit val localTimeSetter: Setter[java.time.LocalTime]           =
+      sqlTimeSetter.contramap(java.sql.Time.valueOf)
+    implicit val localDateTimeSetter: Setter[java.time.LocalDateTime]   =
+      sqlTimestampSetter.contramap(java.sql.Timestamp.valueOf)
+    implicit val zonedDateTimeSetter: Setter[java.time.ZonedDateTime]   =
+      forSqlType(
+        (ps, i, value) => ps.setObject(i, value.toOffsetDateTime, Types.TIMESTAMP_WITH_TIMEZONE),
+        Types.TIMESTAMP_WITH_TIMEZONE
+      )
+    implicit val instantSetter: Setter[java.time.Instant]               =
+      forSqlType(
+        (ps, i, value) => ps.setObject(i, OffsetDateTime.ofInstant(value, ZoneOffset.UTC)),
+        Types.TIMESTAMP_WITH_TIMEZONE
+      )
```

Comment:
Note here that I changed the implementation of the `Setter[java.time.Instant]`

JDBC is a chaotic world and it doesn't seem to exist one correct implementation of this `Setter`.

Because the tests are running on PG, I implemented the `Setter[java.time.Instant]` which is working on PG and is not necessarily working on other DBs.
In Quill for example, for MySQL (and only for MySQL), the `java.sql.Types` used is `Types.TIMESTAMP` instead of `Types.TIMESTAMP_WITH_TIMEZONE`. 
See https://github.com/zio/zio-quill/blob/9e9c706f4e6a83a7a655bd64b23a1ade15a145df/quill-jdbc/src/main/scala/io/getquill/context/jdbc/SimplifiedContexts.scala#L48)

Quill also implemented this encoder this way. 
See https://github.com/zio/zio-quill/blob/master/quill-jdbc/src/main/scala/io/getquill/context/jdbc/Encoders.scala#L127-L128

Doobie also implemented this encoder this way. 
See https://github.com/tpolecat/doobie/blob/e79bab24b2ae1bc639f3a846dcdd74343696a786/modules/postgres/src/main/scala/doobie/postgres/JavaTimeInstances.scala#L21-L37

See also `pgjdbc` documentation relative to the mapping `java.time.* types <-> PG types`: https://jdbc.postgresql.org/documentation/query/#using-java-8-date-and-time-classes
There's no support for `Instant`, only `OffsetDateTime` which must be backed by a `TIMESTAMP WITH TIME ZONE` according to the documentation

Reply:
I am not at the PC, but play/anorm uses the  no offset type. Imho the important thing is, if you write and read, the value of the instant needs to be the same independent of the system time. I'd write tests for that for the most common SQL distributions using a test container.